src/transform/src/reduction.rs

```diff
@@ -499,15 +499,11 @@ impl FoldConstants {
         for (input_row, diff) in rows {
             let datums = input_row.unpack();
             let temp_storage = RowArena::new();
-            let mut output_rows = func
-                .eval(
-                    exprs
-                        .iter()
-                        .map(|expr| expr.eval(&datums, &temp_storage))
-                        .collect::<Result<Vec<_>, _>>()?,
-                    &temp_storage,
-                )?
-                .fuse();
+            let datums = exprs
```

Comment:
Out of curiosity: Would it be possible to make the evaluation of `datums` lazy?

Reply:
I think this is the same code just reformatted, right? There are other places we do something similar-ish, by explicitly determining the columns in the support of `func` and only producing those. It's a bit lazier, but e.g. doesn't do the conditional evaluation that `If` clauses could allow.

Reply:
Or, in some sense `exprs` is a lot like a `MapFilterProject` and we could apply a projection to it, if we had that information. Of course, I might misunderstand the intent of making things lazy.

Reply:
Yeah, same code, but extracted into a variable so that the call to `eval` can borrow it, which is new. It does seem like it'd be possible for `eval` to take an iterator instead of a slice, but inclined to defer that to another PR to avoid ballooning the scope here.